### PR TITLE
Add documentation

### PR DESCRIPTION
**Pull request by lisajulia**
This PR builds on #4029 which should be merged first.
It adds documentation to some methods that are 'not-so-clear'.

### CODE REVIEW

**Comment by lisajulia** [comment]
jenkins build this please

**Comment by lisajulia** [comment]
jenkins build this please

**Comment by lisajulia** [comment]
jenkins build this please

**Comment by lisajulia** [comment]
jenkins build this please

**Comment by lisajulia** [comment]
jenkins build this please

**Comment by blattms** [comment]
Just one question. With these changes, do we need to update the *.pyi file?

**Comment by lisajulia** [comment]
jenkins build this please

**Comment by lisajulia** [comment]
jenkins build this please

**Comment by lisajulia** [comment]
jenkins build this please

**Comment by lisajulia** [comment]
jenkins build this please

File: CMakeLists.txt [hunk 1]
```diff
@@ -157,6 +157,18 @@ macro (sources_hook)
...
+  set(PYTHON_DOCSTRINGS_GENERATED_HPP "${PROJECT_BINARY_DIR}/python/cxx/OpmCommonPythonDoc.hpp")
+  # Command to run the Python script
+  add_custom_command(
+      OUTPUT ${PYTHON_DOCSTRINGS_GENERATED_HPP}
+      COMMAND ${PYTHON_EXECUTABLE} ${PROJECT_SOURCE_DIR}/python/generate_docstring_hpp.py
+              ${PYTHON_DOCSTRINGS_FILE} ${PYTHON_DOCSTRINGS_GENERATED_HPP} OPMCOMMONPYTHONDOC_HPP "Opm::Common::DocStrings"
+      DEPENDS ${PYTHON_DOCSTRINGS_FILE}
+      COMMENT "Generating OpmCommonPythonDoc.hpp from JSON file"
+  )
+  list(INSERT opm-common_SOURCES 0 ${PYTHON_DOCSTRINGS_GENERATED_HPP})
```
Comment (bska) [inline CMakeLists.txt:171]: Does this run unconditionally&ndash;i.e., whether Python is found/enabled or not?

**Comment by bska** [comment]
jenkins build this opm-simulators=5336 please

**Comment by lisajulia** [comment]
jenkins build this opm-simulators=5336 please

**Comment by lisajulia** [comment]
jenkins build this opm-simulators=5336 please

**Comment by lisajulia** [comment]
jenkins build this opm-simulators=5336 please

**Comment by lisajulia** [comment]
jenkins build this opm-simulators=5336 please

File: CMakeLists.txt (hunk 1 above)
Comment (lisajulia) [reply CMakeLists.txt]: I've changed this now, thanks!

**Comment by lisajulia** [comment]
@hakonhagland, @blattms, @akva2: I've finally found the time to fix the things in this PR and https://github.com/OPM/opm-simulators/pull/5349 - can you have a look?
To make the whole "automatic deployment of the documentation work" we need to add a token here, such that a change in opm-common/python/docstrings_common.json triggers [this workflow](https://github.com/OPM/opm-simulators/pull/5349/files#diff-dd5cc39c9889ef2e245e0be2685d661310eb937124f823fd813889f8ea76d65c) in opm-simulators.

For this we need to:
- As the OPM-user, go to https://github.com/settings/apps
- Create an access token: 
Screenshot from 2024-06-25 17-29-48
- Go to opm-common, add the access-token as the PERSONAL_ACCESS_TOKEN:
Screenshot from 2024-06-25 17-24-55

**Comment by hakonhagland** [comment]
jenkins build this please

**Comment by lisajulia** [comment]
@blattms and @hakonhagland: This is finally ready for review :)

**Comment by lisajulia** [comment]
jenkins build this please

**Comment by lisajulia** [comment]
jenkins build this please

File: .github/workflows/dispatch_opm_simulators.yml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      uses: actions/checkout@v4
+
+    - name: Send dispatch to opm-simulators
+      env:
+        PERSONAL_ACCESS_TOKEN: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
+      run: |
+        curl -X POST \
+          -H "Accept: application/vnd.github+json" \
+          -H "Authorization: token ${{ secrets.PERSONAL_ACCESS_TOKEN }}" \
+          https://api.github.com/repos/lisajulia/opm-simulators/dispatches \
```
Comment (hakonhagland) [inline .github/workflows/dispatch_opm_simulators.yml:28]: I guess this needs to change to `OPM` instead of `lisajulia` when this workflow is activated ?

**Review by hakonhagland — APPROVED** [review]

**Comment by blattms** [comment]
Approved by @hakonhagland and I trust him. Comment can be addressed once someone actually activates github action. Merging.

File: .github/workflows/dispatch_opm_simulators.yml (hunk 2 above)
Comment (lisajulia) [reply .github/workflows/dispatch_opm_simulators.yml:28]: Yes, thanks! I've opened a new PR: https://github.com/OPM/opm-common/pull/4129

**Pull request merged**